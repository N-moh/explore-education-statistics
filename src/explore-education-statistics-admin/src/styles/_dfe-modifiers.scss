// Override standard gds styling

.dfe-align--left-hand-controls {
  margin-right: 0;
}

.dfe-hide-controls {
  .dfe-comment-block {
    display: none;
  }
}

// stylelint-disable-next-line selector-no-qualifying-type
h1.govuk-panel__title {
  color: govuk-colour('white');
}
// Admin navigation

.app-navigation--current-page a:visited {
  border-bottom: 0.3rem solid $govuk-link-colour;
}

.app-navigation--current-page a {
  border-bottom: 0.3rem solid $govuk-link-colour;
}

.dfe-details-no-margin {
  details {
    margin-bottom: 0;
    padding: 0;
  }

  .govuk-details__summary {
    margin-bottom: 0;
  }
}

.dfe-filter-overflow {
  margin-left: -1rem;
  max-height: 600px;
  overflow-x: auto;
  padding-left: 1rem;
}

.dfe-summary-list__key--small {
  width: 15%;
}

.govuk-link.dfe-link--white {
  color: govuk-colour('white');
  &:hover {
    text-decoration: none;
  }
}

// Hide page status tag from admin pages
.dfe-page-editing,
.dfe-page-preview {
  .govuk-tag {
    display: none;
  }
}

.dfe-multiline-content {
  white-space: pre-line;
}

.govuk-details .govuk-tag {
  font-size: 0.8rem;
}
