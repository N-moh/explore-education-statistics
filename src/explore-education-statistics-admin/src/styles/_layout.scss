@import './variables';

.dfe-admin-header {
  border-bottom: 10px solid govuk-colour('red');
}

.dfe-width-container--wide {
  max-width: $dfe-page-width-wide;
}

.dfe-content-overflow {
  overflow: auto;
  padding-bottom: 0.5em;
  width: 100%;

  // Forces visible scrollbar on browser/OS that might hide it by default
  &::-webkit-scrollbar {
    // -webkit-appearance: none;
    background-color: rgba(0, 0, 0, 0.1);
    height: 8px;
  }

  &::-webkit-scrollbar-thumb {
    background-color: rgba(0, 0, 0, 0.4);
    border-radius: 20px;
  }

  table {
    max-width: 100%;
  }
}
