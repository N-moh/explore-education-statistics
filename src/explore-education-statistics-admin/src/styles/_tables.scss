@import '~govuk-frontend/govuk/components/table/table';

/* stylelint-disable scss/at-extend-no-missing-placeholder */
.govuk-table {
  th {
    @extend .govuk-table__header;
  }

  td {
    @extend .govuk-table__cell;
  }

  .govuk-table__header--numeric,
  .govuk-table__cell--numeric {
    text-align: right;
  }

  caption {
    @extend .govuk-table__caption;
  }

  .govuk-table__header--center {
    text-align: center;
  }
}

.govuk-table__row--striped {
  background: govuk-colour('light-grey');
}
