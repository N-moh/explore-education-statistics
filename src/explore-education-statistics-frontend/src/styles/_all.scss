@import './reset';
@import './typography';

// Patterns that aren't in Frontend
@import './patterns/related-items';
@import './patterns/step-by-step-navigation';
@import './patterns/task-list';
