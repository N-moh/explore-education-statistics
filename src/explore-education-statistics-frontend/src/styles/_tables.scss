@import '~govuk-frontend/components/table/table';

/* stylelint-disable scss/at-extend-no-missing-placeholder */
.govuk-table {
  th {
    @extend .govuk-table__header;
  }

  td {
    @extend .govuk-table__cell;
  }

  .govuk-table__header--numeric,
  .govuk-table__cell--numeric {
    padding: govuk-spacing(2);
    text-align: right;
  }

  .govuk-table__header--center {
    text-align: center;
  }

  .govuk-table__header--border-left,
  .govuk-table__cell--border-left {
    border-left: 1px solid $govuk-border-colour;
  }

  caption {
    @extend .govuk-table__caption;
  }
}

.govuk-table__row--striped {
  background: govuk-colour('grey-4');
}
