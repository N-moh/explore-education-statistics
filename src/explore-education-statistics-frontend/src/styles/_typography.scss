@import '~govuk-frontend/core/lists';
@import '~govuk-frontend/core/section-break';
@import '~govuk-frontend/core/typography';
@import '~govuk-frontend/core/links';
@import '~govuk-frontend/helpers/typography';

body {
  @extend %govuk-body-m;
}

a {
  @extend %govuk-link;
}

strong {
  @include govuk-typography-weight-bold;
}

small {
  @extend %govuk-body-s;
}

ul {
  @extend %govuk-list--bullet;
}

ol {
  @extend %govuk-list--number;
}

dt {
  @include govuk-typography-weight-bold;
}

h1 {
  @extend %govuk-heading-xl;
}

h2 {
  @extend %govuk-heading-l;
}

h3 {
  @extend %govuk-heading-m;
}

h4 {
  @extend %govuk-heading-s;
}

hr {
  @extend %govuk-section-break,
    %govuk-section-break--m,
    %govuk-section-break--visible;
}
