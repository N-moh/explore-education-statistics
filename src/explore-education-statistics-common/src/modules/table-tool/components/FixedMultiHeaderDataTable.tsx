import { FullTableMeta } from '@common/modules/full-table/types/fullTable';
import Footnote from '@common/modules/table-tool/components/Footnote';
import React, { forwardRef, ReactNode, Ref, useEffect, useRef } from 'react';
import styles from './FixedMultiHeaderDataTable.module.scss';
import MultiHeaderTable, { RowHeaderType } from './MultiHeaderTable';

const mobileWidth = 1024;

interface Props {
  caption: ReactNode;
  captionId?: string;
  innerRef?: Ref<HTMLElement>;
  columnHeaders: string[][];
  rowHeaders: RowHeaderType[][];
  rowHeaderIsGroup?: boolean[];
  rows: string[][];
  footnotes?: FullTableMeta['footnotes'];
}

const FixedMultiHeaderDataTable = forwardRef<HTMLElement, Props>(
  (props, ref) => {
    const { caption, captionId = 'dataTableCaption', footnotes = [] } = props;

    const containerRef = useRef<HTMLDivElement>(null);
    const mainTableRef = useRef<HTMLTableElement>(null);

    useEffect(() => {
      const listener = () => {
        if (
          containerRef.current &&
          mainTableRef.current &&
          window.innerWidth <= mobileWidth
        ) {
          mainTableRef.current
            .querySelectorAll<HTMLTableHeaderCellElement>('thead td')
            .forEach(el => {
              // eslint-disable-next-line no-param-reassign
              el.style.transform = '';
            });

          mainTableRef.current
            .querySelectorAll<HTMLTableHeaderCellElement>('thead th')
            .forEach(el => {
              // eslint-disable-next-line no-param-reassign
              el.style.transform = '';
            });

          mainTableRef.current
            .querySelectorAll<HTMLTableHeaderCellElement>('tbody th')
            .forEach(el => {
              // eslint-disable-next-line no-param-reassign
              el.style.transform = '';
            });
        }
      };

      window.addEventListener('resize', listener);

      return () => window.removeEventListener('resize', listener);
    }, []);

    return (
      <figure className={styles.figure} ref={ref}>
        <figcaption>{caption}</figcaption>

        <div
          className={styles.container}
          ref={containerRef}
          role="region"
          tabIndex={-1}
          onScroll={event => {
            const { scrollLeft, scrollTop } = event.currentTarget;

            if (mainTableRef.current && window.innerWidth >= mobileWidth) {
              mainTableRef.current
                .querySelectorAll<HTMLTableHeaderCellElement>('thead td')
                .forEach(el => {
                  // eslint-disable-next-line no-param-reassign
                  el.style.transform = `translate(${scrollLeft}px, ${scrollTop}px)`;
                });

              mainTableRef.current
                .querySelectorAll<HTMLTableHeaderCellElement>('thead th')
                .forEach(el => {
                  // eslint-disable-next-line no-param-reassign
                  el.style.transform = `translate(0, ${scrollTop}px)`;
                });

              mainTableRef.current
                .querySelectorAll<HTMLTableHeaderCellElement>('tbody th')
                .forEach(el => {
                  // eslint-disable-next-line no-param-reassign
                  el.style.transform = `translate(${scrollLeft}px, 0)`;
                });
            }
          }}
        >
          <MultiHeaderTable
            {...props}
            ariaLabelledBy={captionId}
            ref={mainTableRef}
          />
        </div>
        {footnotes.length > 0 && (
          <>
            <h3>Footnotes</h3>
            <Footnote content={footnotes} />
          </>
        )}
      </figure>
    );
  },
);

FixedMultiHeaderDataTable.displayName = 'FixedMultiHeaderDataTable';

export default FixedMultiHeaderDataTable;
