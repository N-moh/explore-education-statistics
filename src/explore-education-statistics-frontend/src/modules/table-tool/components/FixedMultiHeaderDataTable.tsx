import throttle from 'lodash/throttle';
import React, { forwardRef, Ref, useEffect, useRef } from 'react';
import DataTableKeys from './DataTableKeys';
import styles from './FixedMultiHeaderDataTable.module.scss';
import MultiHeaderTable from './MultiHeaderTable';

interface Props {
  caption: string;
  captionId?: string;
  innerRef?: Ref<HTMLElement>;
  columnHeaders: string[][];
  rowHeaders: string[][];
  rows: string[][];
}

const FixedMultiHeaderDataTable = forwardRef<HTMLElement, Props>(
  (props, ref) => {
    const { caption, captionId = 'dataTableCaption' } = props;

    const mainTableRef = useRef<HTMLTableElement>(null);
    const headerTableRef = useRef<HTMLTableElement>(null);
    const columnTableRef = useRef<HTMLTableElement>(null);
    const intersectionTableRef = useRef<HTMLTableElement>(null);

    const setStickyElementSizes = throttle(() => {
      if (mainTableRef.current) {
        const mainTableEl = mainTableRef.current;

        const cloneCellHeightWidth = (
          selector: string,
          tableEl: HTMLTableElement,
        ) => {
          const tableCells = tableEl.querySelectorAll<HTMLTableCellElement>(
            selector,
          );

          mainTableEl
            .querySelectorAll<HTMLTableCellElement>(selector)
            .forEach((el, index) => {
              tableCells[index].style.height = `${el.offsetHeight}px`;
              tableCells[index].style.width = `${el.offsetWidth}px`;
            });
        };

        if (headerTableRef.current) {
          headerTableRef.current.style.width = `${mainTableEl.offsetWidth}px`;

          cloneCellHeightWidth('thead th', headerTableRef.current);
        }

        if (columnTableRef.current) {
          cloneCellHeightWidth(
            'thead tr:first-child td:first-child',
            columnTableRef.current,
          );
          cloneCellHeightWidth('tbody th', columnTableRef.current);
        }

        if (intersectionTableRef.current) {
          cloneCellHeightWidth(
            'thead tr:first-child td:first-child',
            intersectionTableRef.current,
          );
        }
      }
    }, 200);

    useEffect(() => {
      setTimeout(() => {
        setStickyElementSizes();

        window.addEventListener('resize', setStickyElementSizes);
      }, 200);

      return () => {
        window.removeEventListener('resize', setStickyElementSizes);
      };
    });

    return (
      <figure className={styles.figure} ref={ref}>
        <figcaption>
          <strong id={captionId}>{caption}</strong>

          <DataTableKeys />
        </figcaption>

        <div
          className={styles.container}
          role="region"
          tabIndex={-1}
          onScroll={event => {
            const { scrollLeft, scrollTop } = event.currentTarget;

            if (headerTableRef.current) {
              headerTableRef.current.style.top = `${scrollTop}px`;
            }

            if (columnTableRef.current) {
              columnTableRef.current.style.left = `${scrollLeft}px`;
            }

            if (intersectionTableRef.current) {
              intersectionTableRef.current.style.top = `${scrollTop}px`;
              intersectionTableRef.current.style.left = `${scrollLeft}px`;
            }
          }}
        >
          <MultiHeaderTable
            {...props}
            ariaLabelledBy={captionId}
            className={styles.intersectionTable}
            ref={intersectionTableRef}
            ariaHidden
            isStickyColumn
            isStickyHeader
          />
          <MultiHeaderTable
            {...props}
            ariaLabelledBy={captionId}
            className={styles.columnTable}
            ref={columnTableRef}
            ariaHidden
            isStickyColumn
          />
          <MultiHeaderTable
            {...props}
            ariaLabelledBy={captionId}
            className={styles.headerTable}
            ref={headerTableRef}
            ariaHidden
            isStickyHeader
          />

          <MultiHeaderTable
            {...props}
            ariaLabelledBy={captionId}
            ref={mainTableRef}
          />
        </div>
      </figure>
    );
  },
);

FixedMultiHeaderDataTable.displayName = 'FixedHeaderGroupedDataTable';

export default FixedMultiHeaderDataTable;
