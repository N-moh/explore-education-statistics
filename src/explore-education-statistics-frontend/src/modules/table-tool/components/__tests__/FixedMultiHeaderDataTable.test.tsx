import React from 'react';
import { render } from 'react-testing-library';
import FixedMultiHeaderDataTable from '../FixedMultiHeaderDataTable';
import styles from '../FixedMultiHeaderDataTable.module.scss';

describe('FixedMultiHeaderDataTable', () => {
  test('renders a sticky header table', () => {
    const { container } = render(
      <FixedMultiHeaderDataTable
        caption="Test table"
        columnHeaders={[
          ['Col group A', 'Col group B'],
          ['Col group C', 'Col group D'],
        ]}
        rowHeaders={[['Row group A', 'Row group B']]}
        rows={[['1', '2', '3', '4'], ['5', '6', '7', '8']]}
      />,
    );

    const stickyHeader = container.getElementsByClassName(
      styles.headerTable,
    )[0];

    expect(stickyHeader).toHaveAttribute('aria-hidden', 'true');

    expect(stickyHeader.querySelectorAll('thead tr')).toHaveLength(2);
    expect(
      stickyHeader.querySelectorAll('thead tr:nth-child(1) th'),
    ).toHaveLength(2);
    expect(
      stickyHeader.querySelectorAll('thead tr:nth-child(2) th'),
    ).toHaveLength(4);

    expect(stickyHeader).toMatchSnapshot();
  });

  test('renders a sticky column table', () => {
    const { container } = render(
      <FixedMultiHeaderDataTable
        caption="Test table"
        columnHeaders={[
          ['Col group A', 'Col group B'],
          ['Col group C', 'Col group D'],
        ]}
        rowHeaders={[['Row group A', 'Row group B']]}
        rows={[['1', '2', '3', '4'], ['5', '6', '7', '8']]}
      />,
    );

    const stickyColumn = container.getElementsByClassName(
      styles.columnTable,
    )[0];

    expect(stickyColumn).toHaveAttribute('aria-hidden', 'true');

    expect(stickyColumn.querySelectorAll('tbody tr')).toHaveLength(2);
    expect(
      stickyColumn.querySelectorAll('tbody tr:nth-child(1) th'),
    ).toHaveLength(1);
    expect(
      stickyColumn.querySelectorAll('tbody tr:nth-child(2) th'),
    ).toHaveLength(1);

    expect(stickyColumn).toMatchSnapshot();
  });

  test('renders a sticky intersection', () => {
    const { container } = render(
      <FixedMultiHeaderDataTable
        caption="Test table"
        columnHeaders={[
          ['Col group A', 'Col group B'],
          ['Col group C', 'Col group D'],
        ]}
        rowHeaders={[['Row group A', 'Row group B']]}
        rows={[['1', '2', '3', '4'], ['5', '6', '7', '8']]}
      />,
    );

    const intersection = container.getElementsByClassName(
      styles.intersectionTable,
    )[0];

    expect(intersection).toHaveAttribute('aria-hidden', 'true');

    expect(intersection.querySelector('thead tr td')).toHaveAttribute(
      'colspan',
      '1',
    );
    expect(intersection.querySelector('thead tr td')).toHaveAttribute(
      'rowspan',
      '2',
    );

    expect(intersection).toMatchSnapshot();
  });

  test('renders underlying table', () => {
    const { container } = render(
      <FixedMultiHeaderDataTable
        caption="Test table"
        columnHeaders={[
          ['Col group A', 'Col group B'],
          ['Col group C', 'Col group D'],
        ]}
        rowHeaders={[['Row group A', 'Row group B']]}
        rows={[['1', '2', '3', '4'], ['5', '6', '7', '8']]}
      />,
    );

    const table = container.querySelector(
      'table:not([aria-hidden])',
    ) as HTMLElement;

    expect(table.querySelectorAll('thead tr')).toHaveLength(2);
    expect(table.querySelectorAll('thead tr:nth-child(1) th')).toHaveLength(2);
    expect(table.querySelectorAll('thead tr:nth-child(2) th')).toHaveLength(4);

    expect(table.querySelectorAll('tbody tr')).toHaveLength(2);
    expect(table.querySelectorAll('tbody tr:nth-child(1) th')).toHaveLength(1);
    expect(table.querySelectorAll('tbody tr:nth-child(2) th')).toHaveLength(1);

    expect(table).toMatchSnapshot();
  });
});
