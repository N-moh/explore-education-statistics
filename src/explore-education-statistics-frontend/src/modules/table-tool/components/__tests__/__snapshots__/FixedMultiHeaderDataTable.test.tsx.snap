// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixedMultiHeaderDataTable renders a sticky column table 1`] = `
<table
  aria-hidden="true"
  aria-labelledby="Test table"
  class="govuk-table columnTable"
>
  <thead>
    <tr>
      <td
        class="borderBottom borderRight"
        colspan="1"
        rowspan="2"
      />
    </tr>
    <tr />
  </thead>
  <tbody>
    <tr
      class="borderBottom"
    >
      <th
        class="govuk-table__cell--numeric borderRight"
        role="row"
        rowspan="1"
      >
        Row group A
      </th>
    </tr>
    <tr
      class="borderBottom"
    >
      <th
        class="govuk-table__cell--numeric borderRight borderBottom"
        role="row"
        rowspan="1"
      >
        Row group B
      </th>
    </tr>
  </tbody>
</table>
`;

exports[`FixedMultiHeaderDataTable renders a sticky header table 1`] = `
<table
  aria-hidden="true"
  aria-labelledby="Test table"
  class="govuk-table headerTable"
>
  <thead>
    <tr>
      <td
        class="borderBottom borderRight"
        colspan="1"
        rowspan="2"
      />
      <th
        class="govuk-table__header--center borderLeft"
        colspan="2"
        scope="colgroup"
      >
        Col group A
      </th>
      <th
        class="govuk-table__header--center borderLeft"
        colspan="2"
        scope="colgroup"
      >
        Col group B
      </th>
    </tr>
    <tr>
      <th
        class="govuk-table__header--numeric borderBottom borderLeft"
        colspan="1"
        scope="col"
      >
        Col group C
      </th>
      <th
        class="govuk-table__header--numeric borderBottom"
        colspan="1"
        scope="col"
      >
        Col group D
      </th>
      <th
        class="govuk-table__header--numeric borderBottom borderLeft"
        colspan="1"
        scope="col"
      >
        Col group C
      </th>
      <th
        class="govuk-table__header--numeric borderBottom"
        colspan="1"
        scope="col"
      >
        Col group D
      </th>
    </tr>
  </thead>
</table>
`;

exports[`FixedMultiHeaderDataTable renders a sticky intersection 1`] = `
<table
  aria-hidden="true"
  aria-labelledby="Test table"
  class="govuk-table intersectionTable"
>
  <thead>
    <tr>
      <td
        class="borderBottom borderRight"
        colspan="1"
        rowspan="2"
      />
    </tr>
    <tr />
  </thead>
</table>
`;

exports[`FixedMultiHeaderDataTable renders underlying table 1`] = `
<table
  aria-labelledby="Test table"
  class="govuk-table"
>
  <thead>
    <tr>
      <td
        class="borderBottom borderRight"
        colspan="1"
        rowspan="2"
      />
      <th
        class="govuk-table__header--center borderLeft"
        colspan="2"
        scope="colgroup"
      >
        Col group A
      </th>
      <th
        class="govuk-table__header--center borderLeft"
        colspan="2"
        scope="colgroup"
      >
        Col group B
      </th>
    </tr>
    <tr>
      <th
        class="govuk-table__header--numeric borderBottom borderLeft"
        colspan="1"
        scope="col"
      >
        Col group C
      </th>
      <th
        class="govuk-table__header--numeric borderBottom"
        colspan="1"
        scope="col"
      >
        Col group D
      </th>
      <th
        class="govuk-table__header--numeric borderBottom borderLeft"
        colspan="1"
        scope="col"
      >
        Col group C
      </th>
      <th
        class="govuk-table__header--numeric borderBottom"
        colspan="1"
        scope="col"
      >
        Col group D
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      class="borderBottom"
    >
      <th
        class="govuk-table__cell--numeric borderRight"
        role="row"
        rowspan="1"
      >
        Row group A
      </th>
      <td
        class="govuk-table__cell--numeric borderLeft"
      >
        1
      </td>
      <td
        class="govuk-table__cell--numeric"
      >
        2
      </td>
      <td
        class="govuk-table__cell--numeric borderLeft"
      >
        3
      </td>
      <td
        class="govuk-table__cell--numeric"
      >
        4
      </td>
    </tr>
    <tr
      class="borderBottom"
    >
      <th
        class="govuk-table__cell--numeric borderRight borderBottom"
        role="row"
        rowspan="1"
      >
        Row group B
      </th>
      <td
        class="govuk-table__cell--numeric borderLeft"
      >
        5
      </td>
      <td
        class="govuk-table__cell--numeric"
      >
        6
      </td>
      <td
        class="govuk-table__cell--numeric borderLeft"
      >
        7
      </td>
      <td
        class="govuk-table__cell--numeric"
      >
        8
      </td>
    </tr>
  </tbody>
</table>
`;
